artichoke-core/src/hash.rs

```diff
@@ -0,0 +1,9 @@
+//! Build hashers and hash values
+
+/// A trait for creating instances of RandomState
+pub trait BuildHasher<B: std::hash::BuildHasher> {
+    /// Concrete error type for errors encountered when outputting hash errors.
+    type Error;
```

Comment:
oh sorry I may have led you down a confusing path here.

Adding a generic parameter to the trait means that `BuildHasher` can be implemented multiple times for a given interpreter with multiple `BuildHasher` type params. This isn't quite the API I was hoping for.

Instead, like we have the associated `Error` type here, implementors of `BuildHasher` should declare exactly 1 `BuildHasher` type they use.

Something like this:

```suggestion
pub trait BuildHasher {
    /// Concrete error type for errors encountered when outputting hash errors.
    type Error;
    
    /// Concrete build hasher type.
    type BuildHasher: core::hash::BuildHasher;
```

This also uses the `BuildHasher` import via `core` so `artichoke-core` continues to be compilable with no-std.

This structure with associated types is used pretty heavily in `artichoke-core`, for example the PRNG in the `Prng` trait:

https://github.com/artichoke/artichoke/blob/1c1702894e6ef237b8d7644c16f6b1bc9ae5f5e7/artichoke-core/src/prng.rs#L10-L11

Reply:
another reason to prefer an associated item here:

If I'm building part of the standard library that wants to generically constrain the Artichoke interpreter to include a hasher, most of the time I don't care what `BuildHasher` is used, just that the trait is implemented.

Having the generic parameter in the trait means it must be named, even if all we do is add `where B: core::hash::BuildHasher`. Using an associated type with that same bound improves the ergonomics a bit.